x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/jwt/JwtSignatureValidatorTests.java

```diff
@@ -0,0 +1,295 @@
+        Mockito.doAnswer(invocation -> {
+            // count the reload events
+            reloadAttemptCounter.getAndIncrement();
+            // grab the listener that is passed to reload method and call the onResponse method to simulate a successful reload
+            ActionListener<JwkSetLoader.JwksAlgs> listener = invocation.getArgument(0);
+            listener.onResponse(new JwkSetLoader.JwksAlgs(Collections.emptyList(), Collections.emptyList()));
+            return null;
+        }).when(jwkSetLoader).reload(any(ActionListener.class));
+
+        // return the same sha signature. simulates the check for change results in no change after attempting reload
+        Mockito.doAnswer(
+            invocation -> new JwkSetLoader.ContentAndJwksAlgs(
+                "myshavalue".getBytes(StandardCharsets.UTF_8),
+                new JwkSetLoader.JwksAlgs(Collections.emptyList(), Collections.emptyList())
+            )
+        ).when(jwkSetLoader).getContentAndJwksAlgs();
+
+        Mockito.doAnswer(invocation -> {
+            validateSignatureAttemptCounter.getAndIncrement();
+            // don't actually attempt signature, but do throw an exception to represent a failure which will cause a reload
+            throw new RuntimeException("boom");
+        }).when(signatureValidator).validateSignature(any(SignedJWT.class), anyList());
+        signatureValidator.validate(randomIdentifier(), signedJWT, primaryListener);
+        assertThat(validateSignatureAttemptCounter.get(), is(1));
+        assertThat(finalSuccessCounter.get(), is(0));
+        assertThat(finalFailureCounter.get(), is(1));
+        assertThat(reloadAttemptCounter.get(), is(1));
+    }
+
+    @SuppressWarnings("unchecked")
+    public void testWorkflowWithFailureThenSuccess() throws Exception {
+        // failures will trigger a reload
+        Mockito.doAnswer(invocation -> {
+            // count the reload events
+            reloadAttemptCounter.getAndIncrement();
+            // grab the listener that is passed to reload method and call the onResponse method to simulate a successful reload
+            ActionListener<JwkSetLoader.JwksAlgs> listener = invocation.getArgument(0);
+            listener.onResponse(new JwkSetLoader.JwksAlgs(Collections.emptyList(), Collections.emptyList()));
+            return null;
+        }).when(jwkSetLoader).reload(any(ActionListener.class));
+
+        // return a different sha signature on subsequent calls. simulates the check that reload resulted in a change after reload
+        Mockito.doAnswer(invocation -> {
+            String version = "before";
+            if (reloadAttemptCounter.get() > 0) {
+                version = "after";
+            }
+            return new JwkSetLoader.ContentAndJwksAlgs(
+                version.getBytes(StandardCharsets.UTF_8),
+                new JwkSetLoader.JwksAlgs(Collections.emptyList(), Collections.emptyList())
+            );
+        }).when(jwkSetLoader).getContentAndJwksAlgs();
```

Comment:
I think we could in some places it would likely help with readability, but in other places I would prefer to be explicit with the counters and which counts behave differently. probably just a style difference.